### ISSUE TITLE
Python3 shebang for 3.1.0

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#2261]
_Originally posted by @bjeurissen in https://github.com/MRtrix3/mrtrix3/issues/2047#issuecomment-763199900_

See #2047 and #2215.

D'Oh. Better to discover such while it's only on `dev` though.

1. We can just revert to `#!/usr/bin/env python3`, despite the recommendations against use of `env`;
2. [This comment](https://blogs.gnome.org/mcatanzaro/2018/02/16/on-python-shebangs/#comment-504) on the page Donald linked earlier suggests using some shell logic within the shebang itself to determine the appropriate interpreter;
3. The recommendation to incorporate into the build system isn't ideal for us given we're now distributing pre-compiled code, i.e. we can't rely on `configure` or `build`. If however we had a script that set script shebangs, then `configure` could use logic of any arbitrary complexity to try to decide this, and those installing from pre-compiled packages could still use it if the default doesn't work on their system.
4. Still not a fan of requiring users to explicitly specify the interpreter each time...
5. Any other options still on the table?

**Comment by daljit46** [comment]
Given that we merged #2958, can we close this?

**Comment by Lestropie** [comment]
#2958 only deals with how a Python script already being executed will behave as it executes other Python scripts. It doesn't modify what will happen when an MRtrix3 Python script is first executed by the user from a shell.

With the CMake-generated Python executables produced in #2920, there is technically the option of hard-coding the Python interpreter of choice into the shebang of those short files; this is how FSL gets their code to be interpreted by their custom Conda environment. However I don't think we want to be doing the same thing ourselves. For one, it wouldn't work for shipped compiled tarballs.

Given [your own prior comment](https://github.com/MRtrix3/mrtrix3/issues/2261#issuecomment-2100332887), I had presumed that you would prefer those CMake-generated executables to change to `#!/usr/bin/env python3`. `3.0.x` runs `#!/usr/bin/env python`, which has become problematic; on `dev` we shifted to `#!/usr/bin/python3` rather than `#!/usr/bin/env python3` based on [this comment](https://github.com/MRtrix3/mrtrix3/pull/2215#issuecomment-718131291), but you had made an argument to the contrary at some point. So I'm quite content to change them all to `#!/usr/bin/env python3`, which should now only require a change in the CMake Python executable generation function.

**Comment by daljit46** [comment]
Following #2958, I was under the (I guess wrong) impression that we wanted to hardcode the Python interpreter *versions* (including minor versions).

> So I'm quite content to change them all to #!/usr/bin/env python3, which should now only require a change in the CMake Python executable generation function.

Yes, that still seems the most sensible option to me.

**Comment by Lestropie** [comment]
> I was under the (I guess wrong) impression that we wanted to hardcode the Python interpreter versions (including minor versions).

I did invest a bit more effort in what you're talking about, but not for the reason you're subsequently inferring.

The Python scripts (and the API more generally) can execute non-*MRtrix3* commands, and those commands have shebangs outside of our control. Given we already had a mechanism for detecting when such a command is written for Python and utilising the same interpreter as that currently executing, I came to the realisation that if that command requests a specific minor version of Python3, and the currently executing interpreter does not match that, then the currently executing interpreter should *not* be used, as the specificity of that shebang may indicate a very specific dependency. So I'm trying to do something more clever in such circumstances, but only for speculative non-*MRtrix3* shebangs, not reflecting an intent for *MRtrix3* shebangs.

Though thinking about it more now, I'm concerned the solution I have is no good. But I'll post that in #2958.
